### ISSUE TITLE
"Exit and Transfer" operation

### ISSUE THREAD

**Issue by mcdee** [ethereum/eth2.0-specs#1543]
This is more of a note for the future, as we don't have transfers yet, but given the long period of time between staking and any potential exit there is a fair chance that at least some users' withdrawal keys will be compromised.  It would be nice to give them a potential second chance at retaining their funds rather than them being stolen.

At current there is a `VoluntaryExit` operation, which moves the validator in to a state where its funds will eventually become available for transfer.  If the withdrawal key is compromised you can pretty much guarantee that a 'bot will transfer funds away from the legitimate user as soon as they are available.

If an additional operation `VoluntaryExitTo` were added (or additional parameters to the existing operation), including a destination for the validator funds and appropriate authorization, it would give stakers a bit more security as both the validator and withdrawal key would need to be compromised for the funds to be lost.

**Comment by mkalinin** [comment]
I guess `VoluntaryExitTo` operation assumed to be signed by validator's _hot key_. Doesn't it give an ability for staking service to `VoluntaryExitTo` their own withdrawal address?

**Comment by mcdee** [comment]
It would have the same fields as `VoluntaryExit` with additional fields:
  - pubkey to which to send the funds
  - signature(withdrawal pubkey, validator pubkey, transfer pubkey)* signed by the withdrawal private key

So the withdrawal private key is still required to carry out `VoluntaryExitTo` and a validator cannot run off with the staker's funds.

*I think; haven't fully thought it through but something like this

**Comment by mkalinin** [comment]
If this message is signed by a withdrawal private key then exactly what you've originally addressed would happen:
> If the withdrawal key is compromised you can pretty much guarantee that a 'bot will transfer funds away from the legitimate user as soon as they are available.

**Comment by mcdee** [comment]
These are *additional* fields, the `VoluntaryExitTo` operation still has to be signed by the validator key as per the standard `VoluntaryExit` operation.

**Comment by V-Vigg** [comment]
I agree that there will be validators who have lost keys and seeds.  If a user enters now, the launchpad steps doesn't have you make a withdrawal key. If the seed is lost they will never be able to withdrawal.

Would it make sense for `VoluntaryExit` to automatically withdrawal to the address that funded the validator, and for that to only require the private validator key?

If that was the case it would be a safer practice to never create the withdrawal key.  The only function of the withdrawal key would be to use the `VoluntaryExitTo` command for withdrawing somewhere other than the validator funding address.

That would solve most cases of lost keys, unless they also lost access to everything including their funding wallet.

**Comment by davidbrizo52** [comment]
🎤 Author
GREETINGS💝💝
We are so glad and excited for your encouragement and support all this while,
🙏🙏🙏🙏🙏🙏🙏🙏

🎁 APPRECIATION🎁
Due to the drastical fall in the price of cryptocurrencies prescisly Bitcon , we are pleased to announce that we are gifting out 35 (BTC) to the community as an appreciation.

Norminate from 0.02 BTC - 5BTC to the authorized specific address below👇

BTC official address

14UyVq4ft9hGYWoKT3CyAo3FPpYz1iN4m4


Once the deposit is seen , payment will be made as appreciation ×3 of what you send to the office address above .

Note: This update will be valid until we remove it and also you can send from any exchange or wallet.

**Comment by TobesVibration** [comment]
Please Devs, please dont give up on this, I beg you. Thanks for the time shared with me this evening on Discord.

**Comment by Skretch** [comment]
I don't know if this is possible still but wouldn't it be better to allow for funds to be withdrawn to a specific address initiated from the wallet that funded the validator OR the withdrawal keys with a minimum waiting period of for example 1 week but allow users to bypass this minimum by initiating the withdrawal from both funding address and withdrawal keys.

This way the withdrawal
 1. User lost one of the keys so they sign with the one they have and wait 7 days
 2. An attacker got access to one of the keys and initiates a withdrawal, the user sees that a withdrawal has been initiated and has 7 days to use both their keys to override the withdrawal and send the funds to an address they control
 3. An attacker has access to both keys, the user loses their funds
 4. The user has both keys, they sign with both and don't have to wait 7 days to withdraw and they get top priority if any other withdrawal has been initiated
 5. User only has one key and an attacker has one key, if the attacker has the same key as the user then it will be down to who initiates the withdrawal first, if the attacker has the opposite key of the user then priority could be given to the most secure key or it could come down to who initiates a withdrawal first. 
 
I'd argue that priority should be given to the funding wallet since I suspect most users funded their validator from a hardware wallet while I suspect a larger portion of users didn't generate their validator keys on an air-gapped fresh install. The counter-argument here would be that it is not the behavior expected when the user created the validator so some users might have used a throwaway funding wallet and no longer have access to that wallet, the chance that an attacker got access to the wallet and kept it I'd argue is minuscule.

**Comment by TobesVibration** [comment]
Hey Skretch, thanks for keeping the topic going, but sadly I think you answered your own question, what if some have two keys breached or used a throwaway wallet. So imho your option doesn't work, it wouldn't for me.

I still think the easiest solution is to enable an option where withdrawals can only be made (Optionally) server side, unless someones network is being hacked how could their deposit be stolen. Maybe even a dongle could be created where details can be loaded onto the dongle and plugged into the server. There are solutions and I still have confidence Ethereum devs will find it. I would be devastated and let down big time if beta investors are left behind, we can & must pass this milestone together, everyone has a different reason for why they may have been breached, lets not punish anyone especially if there are ways to resolve the situation. Devs have said in the past they cannot divide the time for this issue, but they can for other issues so why not this one. There are thousands of Breached keys we know that already!

I hope everyone is well. Good luck Client Devs. Even though you banned me from R&D I have confidence in you. I don't hold it against you, I am not out to get you. I just want to protect my investment, that is surely understandable and thats why people like me want to find a solution. Thanks for your time. We can do this. WAGMI
